primitives/core/src/traits.rs

```diff
@@ -20,29 +20,51 @@
 use std::{
 	borrow::Cow,
 	fmt::{Debug, Display},
-	panic::UnwindSafe,
 };
 
 pub use sp_externalities::{Externalities, ExternalitiesExt};
 
+/// Default num of pages for the heap is 64 MB.
+const INIT_HEAP_PAGES: u64 = 1024;
+
+/// Maximum is 3 GB worth of heap.
+const MAX_HEAP_PAGES: u64 = 3 * 1024 * 1024 * 1024 / (64 * 1024);
+
 /// Code execution engine.
 pub trait CodeExecutor: Sized + Send + Sync + ReadRuntimeVersion + Clone + 'static {
 	/// Externalities error type.
 	type Error: Display + Debug + Send + Sync + 'static;
 
 	/// Call a given method in the runtime. Returns a tuple of the result (either the output data
 	/// or an execution error) together with a `bool`, which is true if native execution was used.
-	fn call<
-		R: codec::Codec + PartialEq,
-		NC: FnOnce() -> Result<R, Box<dyn std::error::Error + Send + Sync>> + UnwindSafe,
-	>(
+	fn call<R: codec::Codec + PartialEq>(
+		&self,
+		ext: &mut dyn Externalities,
+		runtime_code: &RuntimeCode,
+		method: &str,
+		data: &[u8],
+		use_native: bool,
+	) -> (Result<crate::NativeOrEncoded<R>, Self::Error>, bool) {
+		let mut heap_pages = INIT_HEAP_PAGES;
+		loop {
+			let (r, used_native) = self.do_call(ext, heap_pages, runtime_code, method, data, use_native);
+			if r.is_ok() || used_native || heap_pages == MAX_HEAP_PAGES {
```

Comment:
How does this terminate, if `r` is an error not related to heap exhaustion?

Reply:
It doesn't in this code; it'll just keep executing until it gets to `MAX_HEAP_PAGES` in an attempt to make it work and then give up.

We could make it work as you're suggesting, but we'd need to bound `Error` on an additional trait `IsHeapExhausted`, which identified the heap exhaustion error type. I haven't put that in yet (notice it's still marked as `inprogress`).